File: 2015/11/passwd.py
import re

LEN = 8

def p2n(p):
    i = 0
    n = 0
    for c in reversed(p):
        n += (ord(c)-ord('a')) * (26 ** i)
        i += 1
    return n

def n2p(n):
    cs = []
    while n > 0:
        cs.append(chr((n % 26) + ord('a')))
        n /= 26
    p = ''.join(reversed(cs))
    return 'a'*(LEN-len(p)) + p

def straight(p):
    ns = [p2n(n) for n in p]
    for i in range(len(ns)-2):
        a,b,c = ns[i:i+3]
        if b == a+1 and c == b+1:
            return True
    return False

def valid_chars(p):
    return re.search('[iol]', p) == None

def has_pairs(p):
    r = re.compile(r'((\w)\2)')

    first = r.search(p)
    if not first:
        return False

    substr = p[first.span()[1]:]
    while True:
        second = r.search(substr)
        if not second:
            return False
        if first.groups(0) != second.groups(0):
            return True
        substr = substr[second.span()[1]:]

def valid(p):
    return len(p) == LEN \
       and valid_chars(p) \
       and straight(p) \
       and has_pairs(p)

def next(p):
    return n2p(p2n(p)+1)

def next_valid(p):
    assert(len(p) == LEN)
    while True:
        m = re.search('[iol]', p)
        if m:
            c = chr(ord(m.group(0))+1)
            p = p[:m.span()[0]] + c + 'a'*(LEN-m.span()[1])
            if valid(p):
                return p
        p = next(p)
        if valid(p):
            return p


assert(not valid('hijklmmn'))
assert(not valid('abbceffg'))
assert(next_valid('abcdefgh') == 'abcdffaa')
assert(next_valid('ghijklmn') == 'ghjaabcc')

p = next_valid('cqjxjnds')
print 'Part 1:', p
print 'Part 2:', next_valid(p)

File: 2015/11/passwd.hs
import           Data.Char
import           Data.List
import qualified Data.Set as S

len :: Int
len = 8

p2n :: String -> Int
p2n p = sum [n c i | (c, i) <- zip (reverse p) [0..((length p)-1)]]
    where n c i = ((ord c) - (ord 'a')) * (26 ^ i)

n2p :: Int -> String
n2p n = prefix ++ p
    where p = reverse $ go n
          prefix = (take (len-(length p)) . repeat) 'a'
          c m = chr ((m `mod` 26) + (ord 'a'))
          go 0 = ""
          go m = [c m] ++ go (m `div` 26)

window :: Int -> [a] -> [[a]]
window n = foldr (zipWith (:)) (repeat []) . take n . tails

isStraight :: String -> Bool
isStraight p = or [and [b == a+1,
                        c == b+1] | [a,b,c] <- window 3 ns]
    where ns = map p2n [[c] | c <- p]

validChars :: String -> Bool
validChars p = not $ or [elem c p | c <- "iol"]

pairs :: Eq a => [a] -> [a]
pairs [] = []
pairs [_] = []
pairs (x:y:xs) = if x == y
                 then [x] ++ pairs xs
                 else pairs (y:xs)

hasPairs :: String -> Bool
hasPairs p = (S.size s) > 1
    where s = (S.fromAscList . sort . pairs) p

isValid :: String -> Bool
isValid p = and [length p == len,
                 validChars p,
                 isStraight p,
                 hasPairs p]

next :: String -> String
next p = n2p ((p2n p)+1)

replaceInvalid :: String -> String
replaceInvalid p = map repl p
    where repl 'i' = 'j'
          repl 'o' = 'p'
          repl 'l' = 'm'
          repl c = c

nextValid :: String -> String
nextValid p = go $ next p
    where go q
            | isValid q = q
            | otherwise = go $ replaceInvalid (next q)

main :: IO ()
main = do
    let p = nextValid "cqjxjnds"
    putStrLn $ "Part 1: " ++ p
    putStrLn $ "Part 2: " ++ (nextValid p)
